airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py

```diff
@@ -0,0 +1,268 @@
+        end_date = pendulum.now()
+
+        while buffered_start_date <= end_date:
+            params_list = []
+            for period, metrics in self.METRICS_BY_PERIOD.items():
+                params_list.append(
+                    {
+                        "metric": metrics,
+                        "period": [period],
+                        "since": buffered_start_date.to_datetime_string(),
+                        "until": buffered_start_date.add(days=self.days_increment).to_datetime_string(),
+                    }
+                )
+            yield params_list
+            buffered_start_date = buffered_start_date.add(days=self.days_increment)
+
+    def _get_insight_records(self, params):
+        return self._api.account.get_insights(params=params)._queue
+
+
+class MediaAPI(StreamAPI):
+    INVALID_CHILDREN_FIELDS = ["caption", "comments_count", "is_comment_enabled", "like_count", "children"]
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        children_fields = list(set(fields) - set(self.INVALID_CHILDREN_FIELDS))
+        media = self._get_media({"limit": self.result_return_limit}, fields)
+        for record in media:
+            record_data = record.export_all_data()
+            if record_data.get("children"):
+                record_data["children"] = [
+                    self._get_single_record(child_record["id"], children_fields).export_all_data()
+                    for child_record in record.get("children")["data"]
+                ]
+            yield record_data
+
+    def _get_media(self, params, fields) -> Iterator[Any]:
+        """
+        This is necessary because the functions that call this endpoint return
+        a generator, whose calls need decorated with a backoff.
+        """
+        return self._api.account.get_media(params=params, fields=fields)
+
+    def _get_single_record(self, media_id, fields) -> IGMedia:
+        return IGMedia(media_id).api_get(fields=fields)
+
+
+class StoriesAPI(StreamAPI):
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        media = self._get_stories({"limit": self.result_return_limit}, fields)
+        for record in media:
+            record_data = record.export_all_data()
+            yield record_data
+
+    def _get_stories(self, params, fields: list) -> Iterator[Any]:
+        """
+        This is necessary because the functions that call this endpoint return
+        a generator, whose calls need decorated with a backoff.
+        """
+        return self._api.account.get_stories(params=params, fields=fields)
+
+
+class MediaInsightsAPI(MediaAPI):
+    MEDIA_METRICS = ["engagement", "impressions", "reach", "saved"]
+    CAROUSEL_ALBUM_METRICS = ["carousel_album_engagement", "carousel_album_impressions", "carousel_album_reach", "carousel_album_saved"]
+
+    def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
+        media = self._get_media({"limit": self.result_return_limit}, ["media_type"])
```

Comment:
we don't seem to be implementing paging here given the result_limit?

Reply:
No, it has been implemented. We get the `Cursor` object with the specified amount of `Media` (in our case, it is 100). And we begin to iterate over it, and when the `Cursor` reaches the last Media and reads it, then inside the `facebook_businness` `Cursor` is implemented in such a way that it pulls up the next 100 `Media` (if they exist, of course).

Reply:
Got it - this is not obvious from just reading the code - could you leave a comment explaining this? Just copy pasting your GitHub comment is fine